### ISSUE TITLE
Present Awkward strings to Numba as Numba strings

### ISSUE THREAD

**Issue by jpivarski** [scikit-hep/awkward#1917]
### Description of new feature

I feel very certain that I implemented this, but I can't find the implementation anywhere.

Even in [main-v1 src/awkward/_connect/_numba](https://github.com/scikit-hep/awkward/tree/main-v1/src/awkward/_connect/_numba) (assuming it was overlooked in the porting), all of the references to holding the GIL are related to partitions and C++ shared_ptr references, which are no longer things we need to worry about.

When `array[i]` is called for an `array` of strings (ListArrayType whose `__array__` property is `"string"`), it should return a Numba string, namely `nb.types.string`. That's a Python Unicode object, which requires GIL-holding, but returning that instead of an Awkward list array has the advantage that all of the Numba methods will work for its lowered string type.

What I may be remembering is that ArrayBuilder _consumes_ strings. What this feature request is asking for is for Arrays to _produce_ strings. (And same for bytestrings: `__array__` is `"bytestring"`.)

I'd like to assign this to you, @ianna, since you're focusing on the Numba stuff and are already familiar with how Numba deals with strings. It's lower priority than Numba-CUDA, though.

**Comment by jpivarski** [comment]
This is weird: it's already implemented.

```python
>>> import awkward as ak
>>> import numba as nb
>>> @nb.njit
... def capitalize_i(array, i):
...     return array[i].capitalize()
... 
>>> capitalize_i(ak.Array(["hello", "there"]), 0)
'Hello'
>>> capitalize_i(ak.Array(["hello", "there"]), 1)
'There'
```

I don't know how it's implemented because I don't see the implementation anywhere. As a very low-priority item, you or I should try to find out where the implementation is, just to be satisfied that it's okay, and then close this issue.

**Comment by jpivarski** [comment]
The conversation that inspired this is here: https://numba.discourse.group/t/feature-request-about-supporting-arrow-in-numba/1668

**Comment by dlee992** [comment]
@jpivarski , I have contributed several PRs for numba `unicode_type`. Indeed, you are incorrect about the situation of numba `unicode_type`. IIRC, numba `unicode_type` doesn't hold the GIL, you can see that `UnicodeModel` only borrows the underlying char buffer/data from Python Unicode instance, and recreate all metadata and reimplement non-GIL operations for it.

I didn't verify your test case posted above, but I guess it could be because Numba `unicode_type` has done that for u, if u're sure about the `dtype` of your array is numba `string`, which is an alias for `unicode_type` in numba. 

But I could be wrong, if I misundertand your test case. Hope helpful.

**Comment by jpivarski** [comment]
Maybe that's what I remember wrong: I had thought that it went through the Python C API. If it's a GIL-less Model, that's definitely better, though it means that the only string functions that can exist are special Numba implementations.

Earlier today, when I searched for references to GIL-holding, I didn't find it because the UnicodeModel doesn't need the GIL. Okay, but I also searched for (case-insensitive) "uni" and "str" and didn't find the implementation. I would have ground something named "UnicodeModel."

Very strange! Because I really remember implementing it and the current code demonstrably does it. I'd have to trace through it more slowly to find out how it does it—probably jogging my memory in the process—but it's good to know at least that the implementation exists... somewhere...
